README
NAME
    vimsw - Profile switcher for Vim

VERSION
    This document describes vimsw version 0.0.1

SYNOPSIS
        $ vimsw [command] ([argument(s)])

        Commands:
            init                   Initialize the environment
            list                   Show the Vim profiles
            create [profile name]  Create a empty Vim profile
            switch [profile name]  Switch to the specified Vim profile

        Example:
            $ vimsw init           # Initialize
            $ vimsw list           # Show profiles list
            $ vimsw create scott   # Create the vim profile of "scott"
            $ vimsw switch scott   # Switch the vim profile to "scott"

DESCRIPTION
    vimsw is the profile switcher for Vim.

    This application can switch each profiles and create profiles.

DEPENDENCIES
    Test::File (version 1.34 or later)

    Test::Exception (version 0.31 or later)

    Test::MockObject::Extends (version 1.20120301 or later)

BUGS AND LIMITATIONS
    No bugs have been reported.

AUTHOR
    moznion "<moznion@gmail.com>"

LICENCE AND COPYRIGHT
    Copyright (c) 2013, moznion "<moznion@gmail.com>". All rights reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for App-VimSw

0.0.1  Sat Mar 23 01:06:31 2013
       Initial release.



lib/App/VimSw.pm
package App::VimSw;

use strict;
use warnings;
use utf8;
use Carp;
use Cwd qw/abs_path/;
use File::Copy::Recursive qw/rmove/;
use File::Spec::Functions qw/catfile/;
use File::Path;

our $VERSION = '0.0.1';

sub new {
    my ( $class, $vimsw_dir, $env ) = @_;
    bless {
        vimsw_dir     => $vimsw_dir,
        vimsw_profile => catfile( $vimsw_dir, '.vimsw_profile' ),
        env           => $env,
    }, $class;
}

sub run {
    my ( $self, @args ) = @_;

    my $command = shift @args;
    my $result = eval { $self->$command(@args) };
    croak "$@" if $@;    # FIXME change error message

    return $result;
}

sub init {
    my ($self) = @_;
    my $profile = 'default';

    print "Initializing...\n";

    # Make root directory for VimSw
    my $status = mkpath( catfile( $self->{vimsw_dir}, $profile ) );
    unless ($status) {
        die "Already initialized.\n";
    }

    $self->_swap_symlink_for_entity( '.vim',   $profile );
    $self->_swap_symlink_for_entity( '.vimrc', $profile );

    $self->_rewrite_profile_file($profile);

    print "Initialized!\n";
}

sub list {
    my ($self) = @_;

    my $profiles = $self->_fetch_profiles_list( $self->{vimsw_dir} );

    open( my $fh, '<', $self->{vimsw_profile} )
      or die "Cannot open '$self->{vimsw_profile}': $!\n";
    my $current = readline($fh);
    close($fh);

    chomp($current);
    @$profiles = sort @$profiles;
    foreach my $profile (@$profiles) {
        if ( $profile eq $current ) {
            print '=>';
        }
        else {
            print '  ';
        }
        print "$profile\n";
    }
    print "\n=> : Current Profile\n";

    return $profiles;
}

sub switch {
    my ( $self, $profile ) = @_;

    unless ($profile) {    # Duplicated
        die "Please specify the profile name\n"
          . "Usage: \$ vimsw switch [profile name]\n";
    }

    my $profiles = $self->_fetch_profiles_list( $self->{vimsw_dir} );
    my %profiles;
    $profiles{$_} = 1 for @$profiles;
    unless ( defined $profiles{$profile} ) {
        die "'$profile' does not exist.\n";
    }

    # Rewrite current profile state
    open( my $fh, '>', $self->{vimsw_profile} );
    print $fh $profile;
    close($fh);

    my $home_vim   = catfile( $self->{env}->{HOME}, '.vim' );
    my $home_vimrc = catfile( $self->{env}->{HOME}, '.vimrc' );
    my $vimsw_vim   = catfile( $self->{vimsw_dir}, $profile, '.vim' );
    my $vimsw_vimrc = catfile( $self->{vimsw_dir}, $profile, '.vimrc' );

    $self->_unlink_already_exests_one($home_vim);
    $self->_unlink_already_exests_one($home_vimrc);

    symlink $vimsw_vim,   $home_vim   or die "$!\n";
    symlink $vimsw_vimrc, $home_vimrc or die "$!\n";
}

sub create {
    my ( $self, $profile ) = @_;

    unless ($profile) {    #Duplicated
        die "Please specify the profile name\n"
          . "Usage: \$ vimsw create [profile name]\n";
    }

    my $profile_dir = catfile( $self->{vimsw_dir}, $profile );

    my $status = mkpath( catfile($profile_dir) );
    unless ($status) {
        die "'$profile' already exists.\n";
    }

    mkpath( catfile( $profile_dir, '.vim' ) );
    open( my $fh, '>', catfile( $profile_dir, '.vimrc' ) );
    close($fh);

    print "Create the profile as '$profile'\n";
}

sub _unlink_already_exests_one {
    my ( $self, $path ) = @_;

    if ( -e $path ) {
        unlink $path or die "$!\n";
    }
}

sub _rewrite_profile_file {
    my ( $self, $profile_name ) = @_;

    open( my $fh, '>', $self->{vimsw_profile} )
      or die "Cannot open '$self->{vimsw_profile}': $!\n";
    print $fh $profile_name;
    close($fh);
}

sub _swap_symlink_for_entity {
    my ( $self, $file, $profile ) = @_;

    my $home_side = catfile( $self->{env}->{HOME}, $file );
    my $vimsw_side = catfile( $self->{vimsw_dir}, $profile, $file );

    if ( -l $home_side ) {
        symlink abs_path($home_side), $vimsw_side or die "$!\n";
        unlink $home_side or die "$!\n";
    }
    else {
        rmove $home_side, $vimsw_side or die "$!\n";   # Entity goes to ".vimsw".
    }

    symlink $vimsw_side, $home_side
      or die $!;   # Make symlink on home directory. It refers the moved entity.
}

sub _update_symlink {
    my ($self) = @_;
}

sub _fetch_profiles_list {
    my ( $self, $dir_location ) = @_;

    opendir my $dh, $dir_location or die "Cannot open $dir_location: $!";
    my @directories =
      grep { $_ !~ /^\.\.?$/ && -d catfile( $dir_location, $_ ) } readdir($dh);
    close($dh);

    return \@directories;
}

1;
__END__

=encoding utf8

=head1 NAME

vimsw - Profile switcher for Vim


=head1 VERSION

This document describes vimsw version 0.0.1


=head1 SYNOPSIS

    $ vimsw [command] ([argument(s)])

    Commands:
        init                   Initialize the environment
        list                   Show the Vim profiles
        create [profile name]  Create a empty Vim profile
        switch [profile name]  Switch to the specified Vim profile

    Example:
        $ vimsw init           # Initialize
        $ vimsw list           # Show profiles list
        $ vimsw create scott   # Create the vim profile of "scott"
        $ vimsw switch scott   # Switch the vim profile to "scott"


=head1 DESCRIPTION

vimsw is the profile switcher for Vim.

This application can switch each profiles and create profiles.


=head1 DEPENDENCIES

Test::File (version 1.34 or later)

Test::Exception (version 0.31 or later)

Test::MockObject::Extends (version 1.20120301 or later)


=head1 BUGS AND LIMITATIONS

No bugs have been reported.


=head1 AUTHOR

moznion  C<< <moznion@gmail.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2013, moznion C<< <moznion@gmail.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    name              => 'App-VimSw',
    module_name       => 'App::VimSw',
    script_files      => ['vimsw'],
    license           => 'perl',
    dist_author       => 'moznion <moznion@gmail.com>',
    dist_version_from => 'lib/App/VimSw.pm',
    build_requires    => {
        'Test::More'                => '0.98',
        'Test::MockObject::Extends' => '1.20120301',
        'Test::Exception'           => '0.31',
        'Test::File'                => '1.34',
        'FindBin'                   => 0,
        'File::Copy::Recursive'     => 0,
        'File::Path'                => 0,
        'File::Spec::Functions'     => 0,
    },
    configure_requires => { 'Module::Build' => '0.40' },
    requires           => {
        perl   => '5.008004',
        'Carp' => 0,
        'Cwd'  => 0,
    },
    no_index   => { 'directory' => ['inc'] },
    meta_merge => {
        'resources' => {
            'license'    => 'http://dev.perl.org/licenses/',
            'repository' => 'git://github.com/moznion/p5-App--VimSw.git',
        },
    },
    test_files => ( -d '.git' || $ENV{RELEASE_TESTING} ) ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    add_to_cleanup       => ['App-VimSw-*'],
    create_license       => 1,
    create_readme        => 1,
);

$builder->create_build_script();
